src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs

```diff
@@ -26,10 +27,11 @@ class JsonMessageSerializer : IMessageSerializer
         {
             this.messageMapper = messageMapper;
 
-            settings ??= new JsonSerializerSettings
+            if (settings == null)
             {
-                TypeNameHandling = TypeNameHandling.Auto
-            };
+                settings = new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Auto };
+                log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'.");
```

Comment:
Ah this is super cool

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2328
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2327

Should we mention those links in the warning as a reference? Or refer to a docs site we own that mentiones those analyzer rules as a possible mitigation strategy?

On another note, the above links also mention the serializer binder being an issue. Have we thought about checking that too?

Reply:
those rules don't really show up when using our newtonsoft json package, so I'm not sure why we would link to them other than as a justification why this is a bad default?

We do not check the binders as we use the default settings in that regards and afaik the default settings are safe.

Reply:
I was thinking about an angle to not just say "it is bad" but actually link to information why it is bad.

Reply:
I wanted to keep the log message a bit on the compact side, and I think it's probably not the best idea to link to external resources from within our log messages. But we could link to our own docs and add more documentation there.

Reply:
> Or refer to a docs site we own that mentiones those analyzer rules as a possible mitigation strategy?

Yep I also suggested that. I agree external links are less preferrable